### ISSUE TITLE
Switch from MSYS 2 MinGW64 to MSYS2 UCRT64

### ISSUE THREAD

**Issue by kayahr** [Automattic/node-canvas#2155]
## Issue or Feature

The pre-built Windows binary of node-canvas currently uses DLLs from MSYS2 MinGW64 which uses the old MSVCRT runtime. node-canvas would be more portable and more compatible to other native libraries like [Sharp](https://www.npmjs.com/package/sharp) when it would use the universal runtime instead. 

See https://learn.microsoft.com/en-us/cpp/porting/upgrade-your-code-to-the-universal-crt?view=msvc-170 and https://www.msys2.org/docs/environments/#msvcrt-vs-ucrt

Switching is pretty much straight forward:

* Use **MSYS2 UCRT64** environment instead of **MSYS2 MinGW64**.
* Install the ucrt64 variants of all needed packages (i.e. **ucrt64/mingw-w64-ucrt-x86_64-cairo** instead of **mingw64/mingw-w64-x86_64-cairo**)
* Replace **/mingw64/** paths in scripts with **/ucrt64/**

I guess a pull request makes no sense here because your build environment for the pre-built windows binaries seems to be manually installed.

**Comment by vito-vanhecke** [comment]
How do you go about doing this exactly? We require both node-canvas and Sharp, but we are not familiar with c++?

**Comment by kayahr** [comment]
@vito-vanhecke I wrote "Switching is pretty much straight forward" but this was meant for the node-canvas authors. For a node-canvas user (especially when you are not familiar with C++ build stuff on Windows) it is not that simple I'm afraid.

**Comment by chearon** [comment]
@kayahr a pull request would be much welcome, I think to the `prebuilds` branch (@zbjornson has been doing them lately)

**Comment by vito-vanhecke** [comment]
Are there plans for these prebuilds? We do require both sharp and node-canvas...

**Comment by vito-vanhecke** [comment]
@zbjornson Are you planning these prebuilds? Or is there a workaround? Not using sharp, not using canvas?

**Comment by lovell** [comment]
@suniltandan It's possible this may relate to https://github.com/libvips/libvips/issues/3275 - the forthcoming sharp v0.32.0 will include the fix if so.

**Comment by acalcutt** [comment]
I am having this issue with https://github.com/maptiler/tileserver-gl . We just got maplibre-native binaries that work with windows, but there is a conflict between sharp and canvas that prevents it from working on windows. sharp 0.32.0 was just released, but does not appear to fix this issue like suggested in the post above.

We have found we can downgrade sharp to 0.30.7 and it works, but I am told the newer versions fail because of the issue mentioned here.

**Comment by lovell** [comment]
@acalcutt Are you compiling your own `canvas` binaries from source using the instructions provided? If so, do you see the exact same`g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed` error?

**Comment by acalcutt** [comment]
We are just using the binary provided by the 2.11.0 npm package, so we are not building our own copy I don't believe.  I am told by someone I am working with that they have made a UCRT64 version of the binary like [the 
](https://github.com/Automattic/node-canvas/pull/2219) and it has worked for them

If sharp is imported before canvas, the error looks like this
![image](https://user-images.githubusercontent.com/3792408/227829906-4daac9e0-5ebe-4717-b524-c6646ebd9db7.png)

If canvas is imported before sharp, the error looks like this
![image](https://user-images.githubusercontent.com/3792408/227830313-b37600f0-0777-43ab-8a07-483faff193fe.png)

If canvas is imported before sharp and we supply custom made UCRT64 node-canvas binary, it runs
![image](https://user-images.githubusercontent.com/3792408/227834565-7b6f3355-95ce-4b25-ac5c-b3ed9b863a43.png)

**Comment by kayahr** [comment]
@tdcosta100 This is the whole point of this issue. Both sharp and canvas are compiled with MSYS2 but they use a different C Runtime. Sharp uses the newer and more universal UCRT64 (Used by MSYS2 UCRT64) while the pre-built canvas binary uses the very old MSVCRT (Used by MSYS2 Mingw64). These runtimes are not compatible and therefor the two libraries cannot be used together. This can be fixed by canvas to switch to UCRT64 as well as I described in this issue. I think it is not a good solution to convince the sharp developers to downgrade their build to the old MSVCRT because UCRT64 is the modern way to go on Windows.

**Comment by tdcosta100** [comment]
My point is switching to MSYS UCRT64 make it more compatible, but doesn't fixes all errors, like @acalcutt demonstrated. But doing this allows at least the coexistence of the two libraries, which was impossible with MSVCRT. I think it's possible to find a build scheme that makes it fully compatible, just don't know how yet (possibly using MXE too, but maybe there's a solution using MSYS2).

**Comment by kayahr** [comment]
I actually don't see a demonstrated problem. I understood that Acalcutt tried running canvas and sharp with the pre-built binaries and it didn't work but with a custom canvas build using UCRT64 it worked. That is also my experience.

**Comment by acalcutt** [comment]
Like @tdcosta100 mentioned, there is still an error when sharp is imported before canvas even with the change to UCRT64 , it looks like the first error
![image](https://user-images.githubusercontent.com/3792408/228295264-f3b50719-fc65-4201-a5ff-6d01a183e8e1.png)

This is somewhat a problem for us, since we need to use the oposite order in linux due to https://github.com/lovell/sharp/issues/371 and imports are hard to change the order of since they need to be top level.

**Comment by chearon** [comment]
Hey all, closing this because I just released 2.11.2 which uses UCRT64. I'm not sure I understand the remaining problems with sharp, but open to anything that works. The only upcoming changes to prebuilds is that we're going to use NAPI and possibly make some of the libraries static.

**Comment by acalcutt** [comment]
Thanks for adding this  :-)

**Comment by octo-kumo** [comment]
For anyone in the future (or the past), I solved the conflict by using sharp@0.31.3 with canvas@2.11.2.

**Comment by jacopomaroli** [comment]
There are two issue at play here.  
1) including/using both (solved)
2) including/using both AND render an svg with text 
Please run the following example **in a cmd terminal** (debugging in vscode won't show errors)

install latest deps:
```
"canvas": "2.11.2",
"sharp": "0.32.3"
```

```
const canvas = require('canvas')
const sharp = require('sharp')

async function test() {
  const c = canvas.createCanvas(200, 200)
  // 1. comment next line to make it work
  const ctx = c.getContext('2d')

  const collectionNameText = Buffer.from(`
    <?xml version="1.0" encoding="UTF-8"?>
    <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200">
      <!-- 2. empty next tag to make it work -->
      <text>hello</text>
    </svg>
  `.trim())

  const buf = await sharp(collectionNameText).toBuffer()
  console.log(buf)
}

test()
```
observe errors:
```
Fontconfig error: Cannot load default config file: No such file: (null)

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.512: cannot register existing type 'GInitable'

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.513: g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed

(process:12092): GLib-CRITICAL **: 19:52:07.514: g_once_init_leave: assertion 'result != 0' failed

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.516: cannot register existing type 'PangoFontMap'

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.517: cannot add private field to invalid (non-instantiatable) type '<invalid>'

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.517: cannot register existing type 'GListModel'

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.518: g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed

(process:12092): GLib-CRITICAL **: 19:52:07.518: g_once_init_leave: assertion 'result != 0' failed

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.518: g_type_add_interface_static: assertion 'G_TYPE_IS_INSTANTIATABLE (instance_type)' failed

(process:12092): GLib-CRITICAL **: 19:52:07.519: g_once_init_leave: assertion 'result != 0' failed

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.519: g_type_register_static: assertion 'parent_type > 0' failed

(process:12092): GLib-GObject-CRITICAL **: 19:52:07.520: cannot add private field to invalid (non-instantiatable) type '<invalid>'
```

Now note the two comments in the above code: following the instructions of either of the two will make it run with no errors.
This was already mentioned here https://github.com/lovell/sharp/issues/2531 and it's likely the issue mentioned above https://github.com/Automattic/node-canvas/issues/2155#issuecomment-1422424184
